test/unity/toolkits/drawing_classifier/test_dc_prediction.cxx

```diff
@@ -0,0 +1,584 @@
+      TS_ASSERT_EQUALS(result_rank.size(), num_of_rows * realk);
+      TS_ASSERT_EQUALS(result_rank["rank"].size() / realk,
+                       expected[PRED_NAME].size());
+      // label should be string type
+      TS_ASSERT_EQUALS(result_rank["rank"][0].get_type(),
+                       flex_type_enum::INTEGER);
+      TS_ASSERT_EQUALS(result_rank["class"][0].get_type(),
+                       flex_type_enum::STRING);
+
+      verify_topk_result(realk, num_of_classes, num_of_rows, result_rank,
+                         expected);
+    };
+
+    prediction_test_driver(batch_size, num_of_rows, num_of_classes, runner);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(test_drawing_classifier_predict_topk_prob_zero_k) {
+  log_for_debug("test_drawing_classifier_predict_topk_prob_zero_k");
+
+  size_t batch_size = 2;
+  size_t num_of_rows = 1;
+  size_t num_of_classes = 2;
+
+  size_t kk = 0;
+  auto runner = [=](test_drawing_classifier& mock_model, gl_sframe my_data,
+                    gl_sframe expected) {
+    auto result_rank = mock_model.predict_topk(my_data, "probability", kk);
+
+    TS_ASSERT_EQUALS(result_rank.size(), num_of_rows);
+    TS_ASSERT_EQUALS(result_rank["probability"].size(),
+                     expected[PRED_NAME].size());
+    // stack empty list will return undefined value
+    TS_ASSERT_EQUALS(result_rank["probability"][0].get_type(),
+                     flex_type_enum::UNDEFINED);
+  };
+
+  prediction_test_driver(batch_size, num_of_rows, num_of_classes, runner);
+}
+
+BOOST_AUTO_TEST_CASE(test_drawing_classifier_predict_topk_prob_normal_k) {
+  log_for_debug("test_drawing_classifier_predict_topk_prob_normal_k");
+
+  for (auto& entry : TEST_CASES) {
+    size_t batch_size = std::get<0>(entry);
+    size_t num_of_rows = std::get<1>(entry);
+    size_t num_of_classes = std::get<2>(entry);
+
+    if (num_of_classes == 0) continue;
+
+    size_t kk = num_of_classes > 1 ? num_of_classes - 1 : 1;
+
+    auto runner = [=](test_drawing_classifier& mock_model, gl_sframe my_data,
+                      gl_sframe expected) {
+      auto result_rank = mock_model.predict_topk(my_data, "probability", kk);
+
+      TS_ASSERT_EQUALS(result_rank.size(), num_of_rows * kk);
+      TS_ASSERT_EQUALS(result_rank["probability"].size() / kk,
+                       expected[PRED_NAME].size());
+      // label should be string type
+      TS_ASSERT_EQUALS(result_rank["probability"][0].get_type(),
+                       flex_type_enum::FLOAT);
+      TS_ASSERT_EQUALS(result_rank["class"][0].get_type(),
+                       flex_type_enum::STRING);
+
+      verify_topk_result(kk, num_of_classes, num_of_rows, result_rank,
+                         expected);
+    };
+
+    prediction_test_driver(batch_size, num_of_rows, num_of_classes, runner);
+  }
+}
+
+BOOST_AUTO_TEST_CASE(test_drawing_classifier_predict_topk_prob_big_k) {
+  log_for_debug("test_drawing_classifier_predict_topk_prob_big_k");
+
+  for (auto& entry : TEST_CASES) {
```

Comment:
I think we can replace `auto` here by `std::tuple<unsigned, unsigned, unsigned>`?

Reply:
no, because I may change the TEST_CASES signature to add properties.

Reply:
besides, for an experienced cpp player, `get<0>` says everything.